### ISSUE TITLE
EAGLE-3 Training and Test Data

### ISSUE THREAD

**Issue by ebubekir-pulat** [SafeAILab/EAGLE#251]
In the newly added code for EAGLE-3 training, there is references to the training data of "/home/lyh/code/nlp/developing/vllmbase/vllm/gedata/l318b.jsonl" and the testing data of "/home/lyh/code/nlp/developing/vllmbase/vllm/gedata/0318.json"?

Do paths refer to the training and test data used in the EAGLE-3 paper? If so, can information be provided on how to access this data, or to extract it ourselves from the relevant datasets?

**Comment by jiahe7ay** [comment]
<img width="1056" alt="Image" src="https://github.com/user-attachments/assets/3c71b180-2062-411e-bdf5-2270a68c84bd" /> By the way, this is the accept  length for Qwen3-8B-Eagle3 in code generation, and its TPS (tokens per second) can reach nearly 500.

**Comment by zxyscz** [comment]
> We have successfully trained the Eagle3 versions of Qwen3-8B and Qwen3-30B-A3B based on the official training code, and have open-sourced them. On a single H200 GPU using the sglang inference framework, Qwen3-8B with Eagle3 achieves a performance boost from 186 tokens/second to 365 tokens/second, while Qwen3-30B-A3B with Eagle3 improves from 147 tokens/second to 231 tokens/second.
> 
> We used the ultra_200k test set and re-ran inference on Qwen3 to regenerate the data, which was then used as the final training set.A total of 600K dialogues were used as the training set.
> 
> https://huggingface.co/Tengyunw/qwen3_30b_moe_eagle3
> 
> https://huggingface.co/Tengyunw/qwen3_8b_eagle3
> 
> Additionally, we have also published a report detailing how to reproduce the Eagle3 training process. The report link is provided below for your reference if needed.
> 
> https://mp.weixin.qq.com/s/Dmdg6aLgFHZEcm6TY1vKkA
> 
> https://zhuanlan.zhihu.com/p/1923763301432662012

During the training process, is config.json replaced with the one for Qwen3?

**Comment by jiahe7ay** [comment]
> > <img alt="Image" width="1056" src="https://private-user-images.githubusercontent.com/33346376/461611963-3c71b180-2062-411e-bdf5-2270a68c84bd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTE2MDM1MjcsIm5iZiI6MTc1MTYwMzIyNywicGF0aCI6Ii8zMzM0NjM3Ni80NjE2MTE5NjMtM2M3MWIxODAtMjA2Mi00MTFlLWJkZjUtMjI3MGE2OGM4NGJkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzA0VDA0MjcwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZlZTBhZDNkNzQ5ZjQwMmE4NDgzZjMxMTBlOGVjMzczNWQyMTE1NjE1MzE1MGJmNzdhYjIxNTJlMTllMGFjYmEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.RqmbYcZyz4q_cNg10ULlhNeCObG0WAVmb-8eIoXlfLU"> [@hongyanz](https://github.com/hongyanz) By the way, this is the accept length for Qwen3-8B-Eagle3 in code generation, and its TPS (tokens per second) can reach nearly 500.
> 
> [@jiahe7ay](https://github.com/jiahe7ay) May I ask which code generation benchmark are you using?

just ulral_chat_200k，but But we regenerated it using the model. We have now open-sourced our training data regenerated using Qwen3-8B.  https://huggingface.co/datasets/Tengyunw/qwen3_8b_eagle3

**Comment by zxyscz** [comment]
@jiahe7ay dou you have this question when inference with eagle3 ? sglang version==0.4.8

<img width="1027" height="159" alt="Image" src="https://github.com/user-attachments/assets/fcdfbff5-d3ac-41f3-8c7c-094ca9253184" />

**Comment by c-dafan** [comment]
I have a question that Qwen3-30B-A3B eagle head_dim is 64, but head_dim of Qwen3-30B-A3B is 128. They are different.

**Comment by c-dafan** [comment]
Hello. I have run this on vLLM with num_spec_tokens=1(draft token=1). When testing with the GSM8K dataset, the accept ratio came out to be 60%. Would you please tell me the accept ratio for Sglang?

**Comment by seohyunwoo-0407** [comment]
@jiahe7ay Thanks for your description, I successed porting EAGLE3 on model "OLMoE-1B-7B-0125-Instruct"! [](https://huggingface.co/wantsleep/OLMoE_1B_7B_Eagle3) Here is draft model of OLMoE! Thank you!

**Comment by fan-niu** [comment]
@jiahe7ay Hi thanks a lot for your works, I use deepspeed zero3, 16384 or 32768 length in llama3-8b 8xh100, batch=1 and still get OOM. How do you avoid OOM problems? thanks. CC: @hongyanz

**Comment by hongyanz** [comment]
@fan-niu We only trained model with 2k context length before.

**Comment by fan-niu** [comment]
@hongyanz @jiahe7ay thanks a lot for those works, when I use the ultra_chat 200k data (without regenerating the assistant data from the target model) to train the llama3.1-8b-instruct model, the training acc is only around 35%. I found that there are errors in the loss mask processing. Is the data processing in the current version of eagle/traineagle3/main.py compatible with llama3.1-8b-instruct? If so, how can I fix this problem? Thank you.

**Comment by fan-niu** [comment]
> We have successfully trained the Eagle3 versions of Qwen3-8B and Qwen3-30B-A3B based on the official training code, and have open-sourced them. On a single H200 GPU using the sglang inference framework, Qwen3-8B with Eagle3 achieves a performance boost from 186 tokens/second to 365 tokens/second, while Qwen3-30B-A3B with Eagle3 improves from 147 tokens/second to 231 tokens/second.
> 
> We used the ultra_200k test set and re-ran inference on Qwen3 to regenerate the data, which was then used as the final training set.A total of 600K dialogues were used as the training set.
> 
> https://huggingface.co/Tengyunw/qwen3_30b_moe_eagle3
> 
> https://huggingface.co/Tengyunw/qwen3_8b_eagle3
> 
> Additionally, we have also published a report detailing how to reproduce the Eagle3 training process. The report link is provided below for your reference if needed.
> 
> https://mp.weixin.qq.com/s/Dmdg6aLgFHZEcm6TY1vKkA
> 
> https://zhuanlan.zhihu.com/p/1923763301432662012

@jiahe7ay Could you please share your code so we can see how you reproduced? Thank you very much. I've read your article on Zhihu and the changes only in dataprocess and load_model (and possibly the hidden idx). Are these the only changes you made? Thanks a again.
